contextualized/regression/datasets.py

```diff
@@ -1,121 +1,120 @@
         self.C = torch.tensor(C, dtype=dtype)
         self.X = torch.tensor(X, dtype=dtype)
         self.Y = torch.tensor(Y, dtype=dtype)
-        self.n_i = 0
-        self.x_i = 0
-        self.y_i = 0
-        self.n = C.shape[0]
         self.c_dim = C.shape[-1]
         self.x_dim = X.shape[-1]
         self.y_dim = Y.shape[-1]
+        self.n = self.C.shape[0]
         self.dtype = dtype
 
-    def __iter__(self):
+        if idx == []:  # use full dataset
+            self.idx = range(self.n)
+        else:
+            self.idx = idx
+
+        self._generate_samples()
+
+    def _generate_samples(self):
         self.n_i = 0
-        self.x_i = 0
         self.y_i = 0
-        return self
+        self.x_i = 0
+        self.all_samples = []
+
+        for i in range(self._total_len()):
+            self.all_samples.append(self._get_next_item(i))
+
+        self.samples = [self.all_samples[i] for i in self.idx]
 
     @abstractmethod
-    def __next__(self):
+    def __len__(self):
         pass
 
     @abstractmethod
-    def __len__(self):
+    def __getitem__(self):
         pass
 
 
-class MultivariateDataset(Dataset):
+class UnivariateDataset(RegressionDatasetBase):
     """
-    Simple multivariate dataset with context, predictors, and outcomes.
+    Simple univariate dataset with context, predictors, and one outcome.
     """
 
-    def __next__(self):
+    def _get_next_item(self, i):
         if self.n_i >= self.n:
             self.n_i = 0
             raise StopIteration
-        ret = (
-            self.C[self.n_i],
-            self.X[self.n_i].expand(self.y_dim, -1),
-            self.Y[self.n_i].unsqueeze(-1),
-            self.n_i,
-        )
+        ret = self.n_i
         self.n_i += 1
         return ret
 
-    def __len__(self):
+    def _total_len(self):
         return self.n
 
+    def __len__(self):
+        return len(self.samples)
+
+    def __getitem__(self, idx):
+        n_i = self.samples[idx]
+        ret = (
+            self.C[n_i],
+            self.X[n_i].expand(self.y_dim, -1).unsqueeze(-1),
+            self.Y[n_i].expand(self.x_dim, -1).T.unsqueeze(-1),
+            n_i,
+        )
+        return ret
+
 
-class UnivariateDataset(Dataset):
+class MultivariateDataset(RegressionDatasetBase):
     """
-    Simple univariate dataset with context, predictors, and one outcome.
+    Simple multivariate dataset with context, predictors, and outcomes.
     """
 
-    def __next__(self):
+    def _get_next_item(self, i):
         if self.n_i >= self.n:
             self.n_i = 0
             raise StopIteration
-        ret = (
-            self.C[self.n_i],
-            self.X[self.n_i].expand(self.y_dim, -1).unsqueeze(-1),
-            self.Y[self.n_i].expand(self.x_dim, -1).T.unsqueeze(-1),
-            self.n_i,
-        )
+        ret = self.n_i
         self.n_i += 1
         return ret
 
-    def __len__(self):
+    def _total_len(self):
         return self.n
 
+    def __len__(self):
```

Comment:
Is there ever a case where `__len__` and `__total_len` will return different values? Remove `_total_len` if not.

Reply:
Yes, sample_ids here would only be the subset of all_sample_ids (for instance in test or train), while total returns all_sample_ids (full dataset)

Reply:
It looks like all_sample_ids is never used outside of constructing sample_ids though. If we have a dataloader constructed as train or test, I don't think there's a reason to keep a list of all sample ids in that object.